tests/xrt/56_xclbin/main.cpp

```diff
@@ -0,0 +1,236 @@
+
+  return 0;
+}
+
+int
+run(int argc, char** argv)
+{
+  if (argc < 3) {
+    usage();
+    return 1;
+  }
+
+  std::string xclbin_fnm;
+  bool verbose = false;
+  unsigned int device_index = 0;
+
+  std::vector<std::string> args(argv+1,argv+argc);
+  std::string cur;
+  for (auto& arg : args) {
+    if (arg == "-h") {
+      usage();
+      return 1;
+    }
+    else if (arg == "-v") {
+      verbose = true;
+      continue;
+    }
+
+    if (arg[0] == '-') {
+      cur = arg;
+      continue;
+    }
+
+    if (cur == "-k")
+      xclbin_fnm = arg;
+    else if (cur == "-d")
+      device_index = std::stoi(arg);
+    else
+      throw std::runtime_error("Unknown option value " + cur + " " + arg);
+  }
+
+  if (xclbin_fnm.empty())
+    throw std::runtime_error("FAILED_TEST\nNo xclbin specified");
+
+  if (device_index >= xclProbe())
+    throw std::runtime_error("Cannot find device index (" + std::to_string(device_index) + ") specified");
+
+
+  auto xclbin = xrt::xclbin(xclbin_fnm); // C++ API, construct xclbin object from filename
+  
+  std::cout << "Targeting xsa platform: " << xclbin.get_xsa_name() << std::endl;
+
+  std::cout << "Xclbin has uuid: " << xclbin.get_uuid() << std::endl;
+
+  std::vector<std::string> cuNames = xclbin.get_cu_names();
+ 
+  for (auto& cuName : cuNames)
+    std::cout << cuName << " ";
+  std::cout << std::endl;
+
+  if (cuNames[0] != "simple:simple_1")
+    throw std::runtime_error("FAILED_TEST\nCould not read correct kernel name, expected: simple:simple_1");
+
+  std::vector<char> data = xclbin.get_data();
+  for(int i=0;i<7;++i)
+    std::cout << data[i] << " ";
+  std::cout << std::endl;
+
+  auto xclbinHandle = xrtXclbinAllocFilename(xclbin_fnm.data()); // C API, construct xclbin object from filename
+
+  if (xclbinHandle == nullptr)
+    throw std::runtime_error("FAILED_TEST\nxrtXclbinAllocFilename returned nullptr.");
+
+  int xsaSz;
+  if (xrtXclbinGetXSAName(xclbinHandle,NULL,&xsaSz) != C_SUCCESS)
+    throw std::runtime_error("FAILED_TEST\nxrtXclbinGetXSAName returned non-zero error code.");
+  
+  char* xsaName = (char*) malloc(xsaSz*sizeof(char));
+  if (xrtXclbinGetXSAName(xclbinHandle,xsaName,&xsaSz) != C_SUCCESS)
+    throw std::runtime_error("FAILED_TEST\nxrtXclbinGetXSAName returned non-zero error code.");
+  std::cout << "Targeting xsa platform: " << xsaName << std::endl;
+  free(xsaName);
+
+  xuid_t c_uuid;
+  if (xrtXclbinGetUUID(xclbinHandle,c_uuid) != C_SUCCESS)
+    throw std::runtime_error("FAILED_TEST\nxrtXclbinGetUUID returned non-zero error code.");
+
+  char** c_cuNames;
+  int numNames;
+  if (xrtXclbinGetCUNames(xclbinHandle,NULL,&numNames) != C_SUCCESS)
+    throw std::runtime_error("FAILED_TEST\nxrtXclbinGetCUNames returned non-zero error code.");
+
+  c_cuNames = (char**) malloc(numNames*sizeof(char*));
+
+  for (int i=0;i<numNames;++i)
+    c_cuNames[i] = (char*) malloc(64*sizeof(char)); // 64 is max size of this field
```

Comment:
ahh, we need to figure a better way for communicate the string name size, it's not part of xrtXclbinGetCUNames so it could write well beyond this (well not since 64 chars is all there is in xclbin, but user could allocate much less per string).  This is just not a good interface.

Also, I'd prefer all xrt test cases to use just the C++ APIs.   We want to de-emphasize the C-APIs, but I realize they should be tested, and I don't have a good suggestion for this.

Reply:
Will remove the C APIs from the test.

Reply:
I am vomiting just looking at this code required by the API... ;-)
But on the other hands, the API is still here and the tests are written, so why not keeping them?